src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -4662,11 +4667,31 @@ Type checkIdInternal(JCTree tree,
                     chk.checkForBadAuxiliaryClassAccess(tree.pos(), env, (ClassSymbol)sym);
                     Type ownOuter = owntype.getEnclosingType();
 
-                    // (a) If symbol is a primitive class and its reference projection
-                    // is requested via the .ref notation, then adjust the computed type to
+                    // (a) If symbol is a primitive class and its reference/value projection
+                    // is requested via the .ref/.val notation, then adjust the computed type to
                     // reflect this.
-                    if (owntype.isPrimitiveClass() && tree.hasTag(SELECT) && ((JCFieldAccess) tree).name == names.ref) {
-                        owntype = new ClassType(owntype.getEnclosingType(), owntype.getTypeArguments(), (TypeSymbol)sym, owntype.getMetadata(), Flavor.L_TypeOf_Q);
+                    if (sym.isPrimitiveClass()) {
+                        if (sym.isReferenceFavoringPrimitiveClass()) {
```

Comment:
Shouldn't we simplify this to try and use `asValueType` - probably this suggests there should be an `asReferenceType` too?

Reply:
ATM, we have com.sun.tools.javac.code.Type#referenceProjection()

Again, I will use JDK-8268734 to rationalize the various internal APIs we have - while doing so, also looking at choice of terminology adopted by java.lang.Class  

But I didn't understand the comment about the simplification you are suggesting - can you clarify ?

Reply:
I mean - this code does a bunch of checks - and then creates a new class type with the opposite flavor/polarity. Isn't that what methods like `Type::asValueType` are supposed to do? E.g. if I have a reference-favoring primitive and I see `.val` can't I just get the type I want by calling `asValueType()` on it? If so, should we also have the dual?

Reply:
> I mean - this code does a bunch of checks - and then creates a new class type with the opposite flavor/polarity. Isn't that what methods like `Type::asValueType` are supposed to do? E.g. if I have a reference-favoring primitive and I see `.val` can't I just get the type I want by calling `asValueType()` on it? If so, should we also have the dual?

OK, understand your point and agree that this could avoid some code duplication, Will follow up. Thanks!